File: test/harness/go/analytics/wrappers_test.go
package analytics

import (
	"testing"

	htevents "github.com/ht-sdks/events-sdk-go"
)

func str(s string) *string { return &s }

func f64(v float64) *float64 { return &v }

func enqueueAndRead(t *testing.T, run func(htevents.Client) error) map[string]interface{} {
	t.Helper()
	body, server := mockServer()
	defer server.Close()

	client, err := htevents.NewWithConfig("wk", htevents.Config{
		Endpoint:  server.URL,
		BatchSize: 1,
	})
	if err != nil {
		t.Fatal(err)
	}
	defer func() { _ = client.Close() }()

	if err := run(client); err != nil {
		t.Fatal(err)
	}

	payload := decodeBatch(<-body)
	if len(payload.Batch) != 1 {
		t.Fatalf("expected 1 message, got %d", len(payload.Batch))
	}
	return payload.Batch[0]
}

func asMap(t *testing.T, v interface{}) map[string]interface{} {
	t.Helper()
	m, ok := v.(map[string]interface{})
	if !ok {
		t.Fatalf("expected map, got %T", v)
	}
	return m
}

func TestTrackOrderCompletedEnqueues(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackOrderCompleted(client, "user-1", TrackOrderCompletedV2Payload{
			OrderID: str("ord_1"),
			Total:   f64(2),
		})
	})
	if msg["type"] != "track" {
		t.Fatalf("type = %v", msg["type"])
	}
	if msg["event"] != "Order Completed" {
		t.Fatalf("event = %v", msg["event"])
	}
	if msg["userId"] != "user-1" {
		t.Fatalf("userId = %v", msg["userId"])
	}
	props := asMap(t, msg["properties"])
	if props["orderId"] != "ord_1" {
		t.Fatalf("orderId = %v", props["orderId"])
	}
	if props["total"] != 2.0 {
		t.Fatalf("total = %v", props["total"])
	}
}

func TestLatestAliasMatchesVersionedWrapper(t *testing.T) {
	versioned := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackOrderCompletedV2(client, "user-1", TrackOrderCompletedV2Payload{
			OrderID: str("1"),
		})
	})
	aliased := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackOrderCompleted(client, "user-1", TrackOrderCompletedV2Payload{
			OrderID: str("1"),
		})
	})
	if versioned["event"] != aliased["event"] {
		t.Fatalf("event mismatch: %v vs %v", versioned["event"], aliased["event"])
	}
}

func TestTrackInjectsContextSchemaVersion(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackSignedUp(client, "user-1", TrackSignedUpDefaultPayload{
			Plan: str("pro"),
		}, CallOptions{
			Context: &htevents.Context{
				Locale: "en-US",
			},
		})
	})
	ctx := asMap(t, msg["context"])
	if ctx["locale"] != "en-US" {
		t.Fatalf("locale = %v", ctx["locale"])
	}
	protocols := asMap(t, ctx["protocols"])
	if protocols["schemaVersion"] != "default" {
		t.Fatalf("schemaVersion = %v", protocols["schemaVersion"])
	}
}

func TestTrackInjectsPropertiesSchemaVersion(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackOrderCompletedPropsV1(client, "user-1", TrackOrderCompletedPropsV1Payload{
			OrderID: str("1"),
		})
	})
	props := asMap(t, msg["properties"])
	if props["orderId"] != "1" {
		t.Fatalf("orderId = %v", props["orderId"])
	}
	if props["apiVersion"] != "v1" {
		t.Fatalf("apiVersion = %v", props["apiVersion"])
	}
}

func TestIdentifyEnqueuesTraits(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return IdentifyDefault(client, "user_1", IdentifyDefaultPayload{
			Email: str("a@b.c"),
		})
	})
	if msg["type"] != "identify" {
		t.Fatalf("type = %v", msg["type"])
	}
	traits := asMap(t, msg["traits"])
	if traits["email"] != "a@b.c" {
		t.Fatalf("email = %v", traits["email"])
	}
}

func TestIdentifyInjectsTraitsSchemaVersion(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return IdentifyTraitsV1(client, "user_1", IdentifyTraitsV1Payload{
			Email: str("a@b.c"),
		})
	})
	traits := asMap(t, msg["traits"])
	if traits["apiVersion"] != "v1" {
		t.Fatalf("apiVersion = %v", traits["apiVersion"])
	}
}

func TestIdentifyDoesNotInjectPropertiesPath(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return IdentifyWrongEnvelopeV1(client, "user_1", IdentifyWrongEnvelopeV1Payload{
			Email: str("a@b.c"),
		})
	})
	traits := asMap(t, msg["traits"])
	if _, ok := traits["apiVersion"]; ok {
		t.Fatalf("did not expect apiVersion on traits: %v", traits)
	}
}

func TestTrackDoesNotInjectTraitsPath(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return TrackWrongEnvelopeV1(client, "user-1", TrackWrongEnvelopeV1Payload{
			OrderID: str("1"),
		})
	})
	props := asMap(t, msg["properties"])
	if _, ok := props["apiVersion"]; ok {
		t.Fatalf("did not expect apiVersion on properties: %v", props)
	}
}

func TestGroupEnqueues(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return GroupDefault(client, "grp_1", "user-1", GroupDefaultPayload{
			Name: str("Acme"),
		})
	})
	if msg["type"] != "group" {
		t.Fatalf("type = %v", msg["type"])
	}
	if msg["groupId"] != "grp_1" {
		t.Fatalf("groupId = %v", msg["groupId"])
	}
	traits := asMap(t, msg["traits"])
	if traits["name"] != "Acme" {
		t.Fatalf("name = %v", traits["name"])
	}
}

func TestPageAndScreenEnqueue(t *testing.T) {
	page := enqueueAndRead(t, func(client htevents.Client) error {
		return PageHome(client, "user-1", PageHomeDefaultPayload{Path: str("/")})
	})
	if page["type"] != "page" {
		t.Fatalf("page type = %v", page["type"])
	}
	if page["name"] != "Home" {
		t.Fatalf("page name = %v", page["name"])
	}

	screen := enqueueAndRead(t, func(client htevents.Client) error {
		return ScreenHome(client, "user-1", ScreenHomeDefaultPayload{Path: str("/")})
	})
	if screen["type"] != "screen" {
		t.Fatalf("screen type = %v", screen["type"])
	}
	if screen["name"] != "Home" {
		t.Fatalf("screen name = %v", screen["name"])
	}
}

func TestAliasEnqueuesWithoutProperties(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return AliasDefault(client, "user_new", "user_old")
	})
	if msg["type"] != "alias" {
		t.Fatalf("type = %v", msg["type"])
	}
	if msg["userId"] != "user_new" {
		t.Fatalf("userId = %v", msg["userId"])
	}
	if msg["previousId"] != "user_old" {
		t.Fatalf("previousId = %v", msg["previousId"])
	}
	if _, ok := msg["properties"]; ok {
		t.Fatalf("alias should not have properties: %v", msg)
	}
}

func TestAliasInjectsContextSchemaVersion(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return AliasContextV1(client, "user_new", "user_old", CallOptions{
			Context: &htevents.Context{Locale: "en-US"},
		})
	})
	if _, ok := msg["properties"]; ok {
		t.Fatalf("alias should not have properties: %v", msg)
	}
	ctx := asMap(t, msg["context"])
	protocols := asMap(t, ctx["protocols"])
	if protocols["schemaVersion"] != "v1" {
		t.Fatalf("schemaVersion = %v", protocols["schemaVersion"])
	}
}

func TestAliasDoesNotInjectPropertiesPath(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		return AliasPropsV1(client, "user_new", "user_old")
	})
	if _, ok := msg["properties"]; ok {
		t.Fatalf("alias should not have properties: %v", msg)
	}
}

func TestCartViewedPreservesJSONKeys(t *testing.T) {
	msg := enqueueAndRead(t, func(client htevents.Client) error {
		itemCount := 3.0
		return TrackCartViewedDefault(client, "user-1", TrackCartViewedDefaultPayload{
			Amount:    10,
			Currency:  "USD",
			ItemCount: &itemCount,
		})
	})
	props := asMap(t, msg["properties"])
	if props["itemCount"] != 3.0 {
		t.Fatalf("itemCount = %v", props["itemCount"])
	}
}

File: test/harness/flutter/test/wrappers_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:hightouch_events/event.dart';

import 'package:analytics_harness/analytics/generated.dart';

import 'support.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('track order completed enqueues name and properties', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackOrderCompleted(
        TrackOrderCompletedV2(orderId: 'ord_1', total: 2.0),
      ),
    );
    expect(track.type, EventType.track);
    expect(track.event, 'Order Completed');
    final props = asMap(track.properties);
    expect(props['orderId'], 'ord_1');
    expect(props['total'], 2);
  });

  test('latest alias matches the versioned wrapper', () async {
    final versioned = await sendAndRead<TrackEvent>(
      (events) => events.trackOrderCompletedV2(
        TrackOrderCompletedV2(orderId: '1'),
      ),
    );
    final aliased = await sendAndRead<TrackEvent>(
      (events) => events.trackOrderCompleted(
        TrackOrderCompletedV2(orderId: '1'),
      ),
    );
    expect(versioned.event, aliased.event);
  });

  test('injects context.protocols.schemaVersion on track', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackSignedUp(
        TrackSignedUpDefault(plan: 'pro'),
        context: {'locale': 'en-US'},
      ),
    );
    final ctx = asMap(track.context!.toJson());
    expect(ctx['locale'], 'en-US');
    expect(schemaVersion(track), 'default');
  });

  test('injects properties.apiVersion on track', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackOrderCompletedPropsV1(
        TrackOrderCompletedPropsV1(orderId: '1'),
      ),
    );
    final props = asMap(track.properties);
    expect(props['orderId'], '1');
    expect(props['apiVersion'], 'v1');
  });

  test('identify posts traits', () async {
    final identify = await sendAndRead<IdentifyEvent>(
      (events) => events.identifyDefault(
        userId: 'user_1',
        traits: IdentifyDefault(email: 'a@b.c'),
      ),
    );
    expect(identify.type, EventType.identify);
    expect(identify.userId, 'user_1');
    expect(asMap(identify.traits!.toJson())['email'], 'a@b.c');
  });

  test('injects traits.apiVersion on identify', () async {
    final identify = await sendAndRead<IdentifyEvent>(
      (events) => events.identifyTraitsV1(
        userId: 'user_1',
        traits: IdentifyTraitsV1(email: 'a@b.c'),
      ),
    );
    expect(asMap(identify.traits!.toJson())['apiVersion'], 'v1');
  });

  test('does not inject properties.* on identify', () async {
    final identify = await sendAndRead<IdentifyEvent>(
      (events) => events.identifyWrongEnvelopeV1(
        userId: 'user_1',
        traits: IdentifyWrongEnvelopeV1(email: 'a@b.c'),
      ),
    );
    expect(asMap(identify.traits!.toJson()).containsKey('apiVersion'), isFalse);
  });

  test('does not inject traits.* on track', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackWrongEnvelopeV1(
        TrackWrongEnvelopeV1(orderId: '1'),
      ),
    );
    expect(asMap(track.properties).containsKey('apiVersion'), isFalse);
  });

  test('group posts groupId and traits', () async {
    final group = await sendAndRead<GroupEvent>(
      (events) => events.groupDefault('grp_1', traits: GroupDefault(name: 'Acme')),
    );
    expect(group.type, EventType.group);
    expect(group.groupId, 'grp_1');
    expect(asMap(group.traits!.toJson())['name'], 'Acme');
  });

  test('page maps to screen and unnamed screen is named screen', () async {
    final page = await sendAndRead<ScreenEvent>(
      (events) => events.pageHome(PageHomeDefault(path: '/')),
    );
    expect(page.type, EventType.screen);
    expect(page.name, 'Home');

    final screen = await sendAndRead<ScreenEvent>(
      (events) => events.screenHome(ScreenHomeDefault(path: '/')),
    );
    expect(screen.type, EventType.screen);
    expect(screen.name, 'Home');

    final unnamed = await sendAndRead<ScreenEvent>(
      (events) => events.screenDefault(ScreenDefault()),
    );
    expect(unnamed.name, 'screen');
  });

  test('alias posts without a properties object', () async {
    final alias = await sendAndRead<AliasEvent>(
      (events) => events.aliasDefault('user_new'),
    );
    expect(alias.type, EventType.alias);
    expect(alias.userId, 'user_new');
  });

  test('injects context.protocols.schemaVersion on alias', () async {
    final alias = await sendAndRead<AliasEvent>(
      (events) => events.aliasContextV1(
        'user_new',
        context: {'locale': 'en-US'},
      ),
    );
    expect(alias.type, EventType.alias);
    expect(schemaVersion(alias), 'v1');
  });

  test('does not inject properties.* on alias', () async {
    final alias = await sendAndRead<AliasEvent>(
      (events) => events.aliasPropsV1('user_new'),
    );
    expect(alias.userId, 'user_new');
    expect(schemaVersion(alias), isNull);
  });

  test('preserves JSON property names on cart viewed', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackCartViewedDefault(
        TrackCartViewedDefault(amount: 10.0, currency: 'USD', itemCount: 3.0),
      ),
    );
    expect(asMap(track.properties)['itemCount'], 3);
  });

  test('json key spellings are preserved on the wire', () async {
    final track = await sendAndRead<TrackEvent>(
      (events) => events.trackJsonKeyProbeDefault(
        TrackJsonKeyProbeDefault.fromMap({
          'order-id': 'hyphen',
          'order_id': 'snake',
          'OrderId': 'pascal',
          'orderId': 'camel',
        }),
      ),
    );
    final props = asMap(track.properties);
    expect(props['order-id'], 'hyphen');
    expect(props['order_id'], 'snake');
    expect(props['OrderId'], 'pascal');
    expect(props['orderId'], 'camel');
  });
}
